### PR TITLE
optimizer: lift more comparisons

### PR DESCRIPTION
**Pull request by aviatesk**
This commit implements more comparison liftings.
Especially, this change enables the compiler to lift `isa`/`isdefined`
checks (i.e. replace a comparison call with ϕ-node by CFG 
union-splitting).

For example, the code snippet below will run 500x faster:
```julia
function compute(n)
    s = 0
    itr = 1:n
    st = iterate(itr)
    while isdefined(st, 2) # mimic our iteration protocol with `isdefined`
        v, st = st
        s += v
        st = iterate(itr, st)
    end
    s
end
```

Although it seems like the codegen for `isa` is fairly optimized already
and so I could not find any performance benefit for `isa`-lifting
(`code_llvm` emits mostly equivalent code), but I hope it's more ideal
if we can do the equivalent optimization on Julia level so that we can
just consult to `code_typed` for performance optimization.

---

@nanosoldier `runbenchmarks("broadcast" || "sparse" || "array" || "union", vs=":master")`

### CODE REVIEW

**Comment by nanosoldier** [comment]
[Your benchmark job](https://github.com/JuliaLang/julia/pull/43227) has completed - possible performance regressions were detected. A full report can be found [here](https://github.com/JuliaCI/NanosoldierReports/blob/master/benchmark/by_hash/0dcf49a_vs_7107ced/report.md).

File: base/compiler/ssair/passes.jl [hunk 1]
```diff
@@ -668,10 +685,14 @@ function sroa_pass!(ir::IRCode)
             canonicalize_typeassert!(compact, idx, stmt)
             continue
         elseif is_known_call(stmt, (===), compact)
-            lift_comparison!(compact, idx, stmt, lifting_cache)
+            lift_comparison!(===, compact, idx, stmt, lifting_cache)
```
Comment (vchuravy) [inline base/compiler/ssair/passes.jl:688]: Do we want to do it for every `no-throw`/`pure` function?

**Review by Keno — APPROVED** [review]

**Comment by aviatesk** [comment]
@nanosoldier `runbenchmarks(!"scalar", vs=":master")`

**Comment by aviatesk** [comment]
This PR seems to increase some allocations:
```
# the latete master
~/julia/packages/EscapeAnalysis master
❯ jld -e 'using EscapeAnalysis; code_escapes(sin, (Int,)); @time code_escapes(Base.init_stdio, (Ptr{Cvoid},)); @time code_escapes(println, (QuoteNode,))'
  6.473641 seconds (24.67 M allocations: 1.275 GiB, 17.72% gc time, 89.63% compilation time)
  0.446880 seconds (2.42 M allocations: 130.121 MiB, 13.46% gc time, 99.98% compilation time)

# before this PR
~/julia/packages/EscapeAnalysis master 18s
❯ ~/julia/julia5/usr/bin/julia -e 'using EscapeAnalysis; code_escapes(sin, (Int,)); @time code_escapes(Base.init_stdio, (Ptr{Cvoid},)); @time code_escapes(println, (QuoteNode,))' 
  4.693671 seconds (24.01 M allocations: 1.261 GiB, 20.25% gc time, 90.90% compilation time)
  0.377396 seconds (2.36 M allocations: 128.876 MiB, 14.03% gc time, 99.98% compilation time)
```

**Comment by nanosoldier** [comment]
[Your benchmark job](https://github.com/JuliaLang/julia/pull/43227#issuecomment-1020870041) has completed - possible performance regressions were detected. A full report can be found [here](https://github.com/JuliaCI/NanosoldierReports/blob/master/benchmark/by_hash/1168642_vs_2682819/report.md).

**Pull request merged**